src/plugins/backend_odbc/backend_odbc_get.c

```diff
@@ -0,0 +1,827 @@
+/**
+ * @file
+ *
+ * @brief Functions for getting data from an ODBC data source.
+ *
+ * This file contains all functions that are especially needed for retrieving data from a data source.
+ * This includes building a string for a SELECT query, preparing and executing select queries and fetching data.
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+
+#include "./backend_odbc_get.h"
+
+#include <kdbassert.h>
+#include <kdberrors.h>
+#include <kdbhelper.h>
+#include <kdblogger.h>
+#include <string.h>
+
+
+/**
+ * @brief Constructs a SELECT query string for the ODBC backend based on the given data source configuration
+ *
+ * @param dsConfig A valid data source config, as returned by the \ref fillDsStructFromDefinitionKs "fillDsStructFromDefinitionKs" function
+ * @param quoteString The characters that should be added before and after identifiers, pass NULL if your identifiers in @dsconfig are
+ * 	already quoted or if you don't want to use quoted identifiers
+ *
+ * @return The query string for the select query to get keynames, key-values and metadata from an SQL data source.
+ * 	Make sure to free the returned string.
+ * @retval "" (empty string) if invalid input detected
+ * 	This string must not be freed!
+ * @retval NULL if memory allocation failed
+ */
+static char * getSelectQueryString (struct dataSourceConfig * dsConfig, char * quoteString)
+{
+	/* A sample query string that shows the structure of the SELECT query that this function generates:
+	 * SELECT "elektra"."key", "elektra"."val", "elektrameta"."metakey", "elektrameta"."metaval" FROM {oj "elektra" LEFT OUTER JOIN
+	 * "elektrameta" ON "elektra"."key"="elektrameta"."key"}
+	 */
+
+	/* Verify that all necessary strings are provided */
+	if (!dsConfig || !(dsConfig->tableName) || !(*(dsConfig->tableName)) || !(dsConfig->keyColName) || !(*(dsConfig->keyColName)) ||
+	    !(dsConfig->valColName) || !(*(dsConfig->valColName)) || !(dsConfig->metaTableName) || !(*(dsConfig->metaTableName)) ||
+	    !(dsConfig->metaTableKeyColName) || !(*(dsConfig->metaTableKeyColName)) || !(dsConfig->metaTableMetaKeyColName) ||
+	    !(*(dsConfig->metaTableMetaKeyColName)) || !(dsConfig->metaTableMetaValColName) || !(*(dsConfig->metaTableMetaValColName)))
+	{
+		return "";
```

Comment:
The idea was to indicate via the return value if a memory allocation failed (`NULL` is returned) or if data to create the query string is missing (empty string is returned).
So, the caller has to check for both cases (NULL or empty) before using (and freeing) the returned value. The situation is also described in the Doxygen comment for the function.

This distinction is used here: https://github.com/ElektraInitiative/libelektra/blob/9adaca1076eea1ae2ad8280c16b81f00b7056f0e/src/plugins/backend_odbc/backend_odbc_get.c#L247-L264


But you have a point here. Maybe it's safer and more flexible to add an argument for passing the `errorKey` to the function `getSelectQueryString` and let it set the error itself, esp. if you consider that the empty string is the only case where a call to free would cause trouble (`free (NULL)` just does nothing).

Reply:
That's likely the best approach.
I've removed the check and added a precondition as you suggested.